extensions/interactions/GraphInput/GraphInput.js

```diff
@@ -187,362 +191,468 @@ oppia.directive('graphViz', function() {
+            $scope.graph.edges[$scope.state.selectedEdge].weight = weight;
+          }
+          return $scope.graph.edges[$scope.state.selectedEdge].weight;
+        };
+
+
+        // Styling functions
+        var DELETE_COLOR = 'red';
+        var HOVER_COLOR = 'aqua';
+        var SELECT_COLOR = 'orange';
+        var DEFAULT_COLOR = 'black';
+        $scope.getEdgeColor = function(index) {
+          if ($scope.state.currentMode === _MODES.DELETE &&
+              index === $scope.state.hoveredEdge &&
+              $scope.canDeleteEdge) {
+            return DELETE_COLOR;
+          } else if (index === $scope.state.hoveredEdge) {
+            return HOVER_COLOR;
+          } else if ($scope.state.selectedEdge === index) {
+            return SELECT_COLOR;
+          } else {
+            return DEFAULT_COLOR;
+          }
+        };
+        $scope.getVertexColor = function(index) {
+          if ($scope.state.currentMode === _MODES.DELETE &&
+              index === $scope.state.hoveredVertex &&
+              $scope.canDeleteVertex) {
+            return DELETE_COLOR;
+          } else if (index === $scope.state.currentlyDraggedVertex) {
+            return HOVER_COLOR;
+          } else if (index === $scope.state.hoveredVertex) {
+            return HOVER_COLOR;
+          } else if ($scope.state.selectedVertex === index) {
+            return SELECT_COLOR;
+          } else {
+            return DEFAULT_COLOR;
+          }
+        };
+        $scope.getDirectedEdgeArrowPoints = function(index) {
+          return graphDetailService.getDirectedEdgeArrowPoints(
+            $scope.graph, index);
+        };
+        $scope.getEdgeCentre = function(index) {
+          return graphDetailService.getEdgeCentre($scope.graph, index);
+        };
+      }]
+  };
+});
+
+oppia.factory('graphInputRulesService', [function() {
+  var isIsomophic = function(graph1, graph2) {
+    if (graph1.vertices.length != graph2.vertices.length) {
+      return false;
+    }
+
+    var constructAdjacencyMatrix = function(graph) {
+      var adjMatrix = [];
+      graph.vertices.map(function() {
+        adjMatrix.push([]);
+      });
+      graph.edges.map(function(edge) {
+        var weight = graph.isWeighted ? edge.weight : 1;
+        adjMatrix[edge.src][edge.dst] = weight;
+        if (!graph.isDirected) {
+          adjMatrix[edge.dst][edge.src] = weight;
         }
-        return $scope.graph.vertices[$scope.state.selectedVertex].label;
-      };
-      $scope.selectedEdgeWeight = function(weight) {
-        if ($scope.state.selectedEdge === null) {
-          return '';
+      });
+      return adjMatrix;
+    };
+
+    var adj1 = constructAdjacencyMatrix(graph1);
+    var adj2 = constructAdjacencyMatrix(graph2);
+
+    var nextPermutation = function(permutation) {
+      // Generates (in place) the next lexicographical permuation.
+      // permutation is a permutation of [0...permutation.length]
```

Comment:
actually it's [0...permutation.length - 1], right?

Also, is it a list of ints? If so, maybe say that explicitly -- I _think_ this is the case, but I'm only getting this sense indirectly, since I'm inferring it from [0...permutation.length]. Maybe say [0, 1, 2, ..., permutation.length - 1]?